### PR TITLE
Developer documentation for Unicode handling

### PR DESCRIPTION
**Pull request by cachedout**

### CODE REVIEW

File: doc/ref/states/writing.rst [hunk 1]
```diff
@@ -269,6 +269,20 @@ logs. The following code snippet demonstrates writing log messages:
...
+have already decoded from strings into Unicode. In Python 2, these will
+be of type 'Unicode' and in Python 3 they will be of type 'bytes'. Calling
+from a state to other Salt sub-systems, such as execution modules should
+pass bytes/Unicode. In the rare event that a state needs to write directly
+to disk, Unicode should be encoded to a string immediatley before writing
+to disk. An author may use ``__salt_system_encoding__`` to learn what the
+encoding type of the system is. For example,
+`'my_string'.encode(__salt_system_encoding__')`.
+
+
```
Comment (s0undt3ch) [inline doc/ref/states/writing.rst]: same comment as above

**Comment by rallytime** [comment]
This is super helpful to help keep this all straight!

**Comment by damon-atkins** [comment]
As part of the transition to Py3 code.     definitions/functions should create local copies of parameters passed and convert them to unicode if they are not unicode already and return unicode in both cases of PY2 and PY3.  This way its building towards PY3.   Then when PY2 support is dropped the code can be simplified over time.  It also assume PY4 will continue with Unicode.


Windows Specific Example
```
def _mbcs_to_unicode(instr):
    if instr is None or isinstance(instr, six.text_type):
        return instr
    else:
        return six.text_type(instr, 'mbcs')

def winexample(hive)
 if PY2:
        local_hive = _mbcs_to_unicode(hive)
    else:
        local_hive = hive
.....

```

**Comment by cachedout** [comment]
I disagree.

Having a hodge-podge of modules, some of which covert Unicode and some of which don't is going to be exceedingly difficult to maintain and will be a mess of code that we'll never be able to clean up.

The better way to do this is to enforce this conversion process at the ingress and egress points and have Unicode everywhere inside the core.

**Comment by damon-atkins** [comment]
> The better way to do this is to enforce this conversion process at the ingress and egress points and have Unicode everywhere inside the core.

I would have thought that is a lot of places.  If that's what SaltStack wants, its needs to be documented.

> ...will be a mess of code that we'll never be able to clean up.
I would have thought, using of six would give you this, which we don't really have a choice about to support PY2 & PY3 at the same time. 

I am seeing code which says if I am PY3 do something different. The code should be if PY2 do something different which should make it easier to pull PY2 support latter and clean up the code. **i.e. the objective should be to make it easy to remove PY2 specific code**, cut of the text in your favourite editor.

The point is Saltstack need to document what they want, so that people can follow.  The above does not provide the guidance for writing PY2 & PY3 code.  Write up a plan so people can get into submitting code to support both.  Maybe it should be in the style guide ??

**Comment by cachedout** [comment]
Go Go Jenkins!

**Pull request merged**